js/private/translate_package_lock.bzl

```diff
@@ -0,0 +1,97 @@
+"Convert package-lock.json into starlark Bazel fetches"
+
+_DOC = """Repository rule to generate npm_import rules from package-lock.json file.
+
+The npm lockfile format includes all the information needed to define npm_import rules,
+including the integrity hash, as calculated by the package manager.
+
+Instead of manually declaring the `npm_imports`, this helper generates an external repository
+containing a helper starlark module `repositories.bzl`, which supplies a loadable macro
+`npm_repositories`. This macro creates an `npm_import` for each package.
+
+Bazel will only fetch the packages which are required for the requested targets to be analyzed.
+Thus it is performant to convert a very large package-lock.json file without concern for
+users needing to fetch many unnecessary packages.
+
+Typical usage:
+```starlark
+load("@aspect_rules_js//js:npm_import.bzl", "translate_package_lock")
+
+# Read the package-lock.json file to automate creation of remaining npm_import rules
+translate_package_lock(
+    name = "npm_deps",
+    package_lock = "//:package-lock.json",
+)
+
+load("@npm_deps//:repositories.bzl", "npm_repositories")
+
+npm_repositories()
+```
+"""
+
+_ATTRS = {
+    "package_lock": attr.label(
+        doc = """The package-lock.json file.
+        
+        It should use the lockfileVersion 2, which is produced from npm 7 or higher.""",
+        mandatory = True,
+    ),
+}
+
+_NPM_IMPORT_TMPL = """    # Generated from [package-lock.json snippet here]
+    npm_import(
+        integrity = "{integrity}",
+        package = "{package}",
+        version = "{version}",
+        deps = {deps},
+    )
+"""
+
+def _escape(package_name):
+    "Make a package name into a valid label without slash or at-sign"
+    return package_name.replace("/", "_").replace("@", "_")
+
+def _repo_name(package_name, version):
+    "Make an external repository name from a package name and a version"
+    return "npm_%s-%s" % (_escape(package_name), version)
+
+def _import_dependencies(repository_ctx, bzl_out, packages):
+    for (name, dep) in packages["dependencies"].items():
+        if "resolved" not in dep.keys():
+            continue
+        deps = []
+        if "requires" in dep.keys():
+            for (n, d) in dep["requires"].items():
+                deps.append("@" + _repo_name(n, d))
+        bzl_out.extend([_NPM_IMPORT_TMPL.format(
+            package = name,
+            version = dep["version"],
+            url = dep["resolved"],
+            integrity = dep["integrity"],
+            deps = deps,
+        )])
+
+def _translate_package_lock(repository_ctx):
+    bzl_content = ["""# Generated by package_lock.bzl from {package_lock}
```

Comment:
Should this have an `@generated` banner?

Reply:
I don't understand, you mean this comment should have different syntax?

Reply:
In addition to, if this is a generated file that's check in, some IDEs and code review tools etc read the `@generated` style comments and treat them differently.